File: lib/network/packets/server/s_inventory.dart
/// 背包裡的一件道具。
///
/// 模板欄位（名稱、類型、可否使用）由伺服器一併帶來，前端因此
/// **不需要一份 item 表的副本** —— 道具定義只存在伺服器，加新道具不必改前端。
class InventoryItem {
  const InventoryItem({
    required this.objId,
    required this.itemId,
    required this.count,
    required this.name,
    required this.itemType,
    required this.stackable,
    required this.usable,
    required this.equipped,
    required this.enchantLevel,
  });

  /// 這一筆的唯一編號。使用／丟棄都以它指定，不是用 itemId ——
  /// 同一種道具可能有多筆（不可疊加的武器）。
  final int objId;

  /// 道具種類編號（→ 伺服器 item 表）。
  final int itemId;
  final int count;
  final String name;

  /// 0=一般 1=武器 2=防具。
  final int itemType;
  final bool stackable;
  final bool usable;
  final bool equipped;
  final int enchantLevel;

  bool get isWeapon => itemType == 1;
  bool get isArmor => itemType == 2;

  /// 顯示名稱：有強化等級就加上 +N。
  String get displayName =>
      enchantLevel > 0 ? '$name +$enchantLevel' : name;

  factory InventoryItem.fromJson(Map<String, dynamic> j) => InventoryItem(
        objId: (j['objId'] as num?)?.toInt() ?? 0,
        itemId: (j['itemId'] as num?)?.toInt() ?? 0,
        count: (j['count'] as num?)?.toInt() ?? 0,
        name: j['name'] as String? ?? '',
        itemType: (j['itemType'] as num?)?.toInt() ?? 0,
        stackable: j['stackable'] as bool? ?? false,
        usable: j['usable'] as bool? ?? false,
        equipped: j['equipped'] as bool? ?? false,
        enchantLevel: (j['enchantLevel'] as num?)?.toInt() ?? 0,
      );
}

/// 整份背包。進遊戲時送一次，之後的變動走 S_ITEM_UPDATE／S_ITEM_REMOVE。
class SInventory {
  const SInventory({required this.items});

  final List<InventoryItem> items;

  factory SInventory.fromData(Map<String, dynamic> data) => SInventory(
        items: [
          for (final i in (data['items'] as List<dynamic>? ?? const []))
            InventoryItem.fromJson(i as Map<String, dynamic>),
        ],
      );
}

/// 單一道具新增或變更。**upsert 語意**：objId 已存在就更新，沒有就新增。
///
/// 所以「撿到新道具」與「數量改變」共用同一個封包。
class SItemUpdate {
  const SItemUpdate({required this.item});

  final InventoryItem item;

  factory SItemUpdate.fromData(Map<String, dynamic> data) => SItemUpdate(
        item: InventoryItem.fromJson(
            data['item'] as Map<String, dynamic>? ?? const {}),
      );
}

/// 道具整筆從背包消失（用光、丟棄、賣掉）。
///
/// 數量減少但還有剩的情況走 S_ITEM_UPDATE —— 前端據此決定是移除該格還是只改數字。
class SItemRemove {
  const SItemRemove({required this.objId});

  final int objId;

  factory SItemRemove.fromData(Map<String, dynamic> data) =>
      SItemRemove(objId: (data['objId'] as num?)?.toInt() ?? 0);
}

File: src/main/java/com/xin/server/packet/server/S_Inventory.java
package com.xin.server.packet.server;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.xin.server.model.instance.ItemInstance;
import com.xin.server.model.instance.PcInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;
import com.xin.server.template.ItemTemplate;

/**
 * 整份背包（對應 opcode {@code S_INVENTORY}）。
 * <p>
 * 選角進遊戲時送一次。之後的變動走 {@link S_ItemUpdate}／{@link S_ItemRemove}
 * 的增量封包，不重送整份 —— 背包會越來越大，每撿一件就整包重送並不划算。
 * <p>
 * 每一件都附上模板欄位（名稱、類型、可否使用…），前端因此<b>不需要一份
 * item 表的副本</b>。道具定義只存在伺服器，前端改版不必跟著更新。
 */
public class S_Inventory extends ServerBasePacket {

    private S_Inventory(PcInstance pc) {
        super(ServerOpcodes.S_INVENTORY);
        ArrayNode arr = newArray();
        for (ItemInstance it : pc.getInventory().getItems()) {
            arr.add(describe(it));
        }
        putArray("items", arr);
        put("size", pc.getInventory().getSize());
    }

    /**
     * 一件道具的完整描述。{@link S_ItemUpdate} 也用同一組欄位 ——
     * 兩邊格式一致，前端只需要一條解析路徑。
     */
    static void fill(ObjectNode n, ItemInstance it) {
        n.put("objId", it.getId());
        n.put("itemId", it.getItemId());
        n.put("count", it.getCount());
        n.put("enchantLevel", it.getEnchantLevel());
        n.put("equipped", it.isEquipped());
        n.put("identified", it.isIdentified());

        ItemTemplate t = it.getItem();
        n.put("name", t != null ? t.getName() : "");
        n.put("itemType", t != null ? t.getItemType() : 0);
        n.put("stackable", t != null && t.isStackable());
        n.put("usable", t != null && t.getUseType() == 1);
        n.put("weaponType", t != null ? t.getWeaponType() : 0);
        n.put("armorType", t != null ? t.getArmorType() : 0);
    }

    private ObjectNode describe(ItemInstance it) {
        ObjectNode n = newObject();
        fill(n, it);
        return n;
    }

    public static S_Inventory of(PcInstance pc) {
        return new S_Inventory(pc);
    }
}
